go/ql/lib/semmle/go/dataflow/internal/TaintTrackingUtil.qll

```diff
@@ -88,13 +98,24 @@ class AdditionalTaintStep extends Unit {
  */
 predicate localAdditionalTaintStep(DataFlow::Node pred, DataFlow::Node succ, string model) {
   (
-    referenceStep(pred, succ) or
-    elementWriteStep(pred, succ) or
-    fieldReadStep(pred, succ) or
-    elementStep(pred, succ) or
-    tupleStep(pred, succ) or
-    stringConcatStep(pred, succ) or
+    referenceStep(pred, succ)
+    or
+    elementWriteStep(pred, succ)
```

Comment:
What's this step doing here? I realise it isn't added in this PR, but it looks wrong: store steps shouldn't be taint steps.

Reply:
The QLDoc says "Holds if there is an assignment of the form `succ[idx] = pred`, meaning that `pred` may taint `succ`." So when you taint an element of an array, you taint the whole array. (And the same for slices, which are just like arrays.)

Reply:
I can see that you could take a different approach, where only the array element is tainted, and then if the array is used without an array access then hopefully there is an implicit read step that will mean that the whole array is tainted. I could try removing this line and see what effect it has. But I think that shouldn't be part of this PR.

Reply:
> I could try removing this line and see what effect it has

I think that should be attempted, yes.

> But I think that shouldn't be part of this PR.

Agree.